Add unsnapped and concurrent object checks

Naxesss opened the pull request:
Makes more progress towards resolving #12091

This adds
- `CheckUnsnappedObjects`, which detects
    - ≥ 2 ms off as a problem
    - ≥ 1 ms off as negligible (this type will eventually be hidden by default)
    - all edges (starts, repeats, ends)
- `CheckConcurrentObjects`, which detects
    - end time being greater than the start time of the previous object
    - per-column in rulesets like mania
- Tests for all the new checks
- `ClosestSnapTime` & `ClosestBeatDivisor` to `IBeatmap`
    - The existing `SnapTime` in `EditorBeatmap` now uses the former
- `IHasColumn` from `osu.Game.Rulesets.Mania` -> `osu.Game`
    - This enables the concurrent check to be ruleset-agnostic

https://user-images.githubusercontent.com/30292137/116154727-521aa200-a6e9-11eb-9dea-bd4a2daea8f3.mp4

---

Because I added a `ClosestBeatDivisor` method for the unsnap check, we're going to have duplicate functionality with the `BeatDivisorFinder` class from https://github.com/ppy/osu/pull/12558. Which approach of these do we prefer? @JustusFT

Naxesss committed:
Add unsnap check

Naxesss committed:
Add unsnap check tests

Naxesss committed:
Move snapping responsibility to `IBeatmap`

Seems `EditorBeatmap` already implements a different kind of `SnapTime` from `IBeatSnapProvider`, so method names here aren't great.

This is very similar to what https://github.com/ppy/osu/pull/12558 is doing, so may need to do some duplicate resolution later, especially surrounding `ClosestBeatSnapDivisor`.

Worth noting that this change makes 1/7, 1/5, etc unsupported for now, as we now rely on `BindableBeatDivisor.VALID_DIVISORS`.

Naxesss committed:
Rename snapping methods

Further separates them from `IBeatSnapProvider`'s `SnapTime`, and groups them together more, to prevent confusion between the two interfaces.

Also changes the xmldoc of the reference time to that of `IBeatSnapProvider` for consistency.

Naxesss committed:
Remove redundant string formatting

Naxesss committed:
Fix duplicate code in unsnap test

Naxesss committed:
Add concurrent objects check

Here we use `IHasColumn` to support rulesets with columns, and so I moved that interface out into `osu.Game` from `osu.Game.Rulesets.Mania`.

We also use the same threshold as the unsnap check to ensure that no problems slip through. Specifically where an object is simultaneously not concurrent and not unsnapped but still on the same tick.

Naxesss committed:
Add concurrent object check tests

Naxesss committed:
Add concurrent objects check to `BeatmapVerifier`

Naxesss committed:
Rename `CheckUnsnaps` -> `CheckUnsnappedObjects`

Will potentially have `CheckUnsnappedKiai` or similar later, so this is worth specifying.

Also consistent with `CheckConcurrentObjects`, which will likely have a `CheckConcurrentLines` later.

Naxesss committed:
Sort objects by time in concurrent check test

Naxesss committed:
Return result of local variable instead

Naxesss committed:
Add snapping time comment

Naxesss committed:
Fix accessibility of `areConcurrent`

Naxesss committed:
Remove unused `virtual` keywords

Added these in a previous iteration, where I had the mania variant inherit this class.

No longer necessary as `IHasColumn` was used to make this check more generic.

Naxesss committed:
Change category of unsnap check to timing

Makes more sense, as this is typically the result of timing changes.

frenzibyte commented on osu.Game/Rulesets/Edit/Checks/CheckUnsnappedObjects.cs:
```diff
@@ -0,0 +1,98 @@
+            }
+        }
+
+        public class IssueTemplate2MsOrMore : IssueTemplateUnsnap
```
nit: going with names like `IssueTemplateLongUnsnap` and `IssueTemplateShortUnsnap` or similar would sound better.

Naxesss committed:
Rename unsnap check templates

JustusFT commented:
> Because I added a ClosestBeatDivisor method for the unsnap check, we're going to have duplicate functionality with the BeatDivisorFinder class from #12558. Which approach of these do we prefer?

I think we can go with your version because it can report the correct divisor more accurately than mine. Mine just gets the first largest divisor within 1ms, so if 2 divisors are within 1ms it might pick the wrong one, but with yours it will pick the closest one. (The BPM would have to be extremely high for that to happen, but still)

Naxesss committed:
Add tests for `ClosestBeatDivisor`

Used https://github.com/ppy/osu/pull/12558/files#diff-5c1f04c5b262ca3abbaf867aa91b62a60b66691323c286ad5aa0b75c153cc6ca as reference.

Naxesss commented:
> Also, we might be able to use the same tests I made for `BeatSnapFinder` to test `ClosestBeatDivisor` (test file)

Yeah, adding it basically verbatim worked. Did some refactoring to cut down on duplicate lines and added another 2 tests, including one for extremely high BPM streams like you mentioned earlier.

peppy committed:
Move snap/divisor helper methods to inside `ControlPointInfo`

peppy commented on osu.Game/Beatmaps/Beatmap.cs:
```diff
@@ -74,6 +75,31 @@ public double GetMostCommonBeatLength()
+            var beatLength = timingPoint.BeatLength / beatDivisor;
+            var beatLengths = (int)Math.Round((time - timingPoint.Time) / beatLength, MidpointRounding.AwayFromZero);
+
+            // Casting to int matches stable.
```
I'm not sure we want this. @smoogipoo opinions? Until now we have been using doubles in the editor and this will force all hitobjects to be snapped to `int` time values.

smoogipoo replied:
I agree, should be using doubles.

peppy committed:
Rename methods to match generally how these find-methods are named elsewhere

peppy committed:
Revert return values to non-rounded doubles

peppy committed:
Refactor lookup methods to avoid linq and reduce `TimingPointAt` calls

peppy commented:
I've made some changes to the location of (moving from `IBeatmap` to `ControlPointInfo`) and efficiency (removed all linq and reduced `TimingPointAt` calls) of the snap lookup methods.

Also made the change back to `double` time values for now. Doesn't look to have broken any of your tests so I think it's fine?

Looks good from my end. @smoogipoo can you confirm whether you're okay with the `IHasColumn` interface being moved out of the osu!mania project?

Naxesss commented:
> Also made the change back to `double` time values for now. Doesn't look to have broken any of your tests so I think it's fine?

Looks to be fine, yeah. Objects in maps made using stable could rarely be treated as unsnapped where they weren't previously, but it's not really a big deal.

Naxesss commented on osu.Game/Rulesets/Edit/Checks/CheckConcurrentObjects.cs:
```diff
@@ -0,0 +1,88 @@
+
+                    // Accounts for rulesets with hitobjects separated by columns, such as Mania.
+                    // In these cases we only care about concurrent objects within the same column.
+                    if ((hitobject as IHasColumn)?.Column != (nextHitobject as IHasColumn)?.Column)
+                        continue;
```
That was my previous implementation for this. Reason I changed was to make it similarly generic to the unsnapped check, which uses `IHasRepeats` instead of having a `CheckUnsnappedReverses`, for example.

No matter what ruleset we're talking about, if it uses columns then this check should apply per-column is the idea, and the current code expresses that much better than splitting out the check across every single ruleset imo.

Any specific reason to have custom rulesets need to implement their own version of `IHasColumn`?

smoogipoo replied:
Hmm, I suppose it's fine... Rulesets can do whatever they want. They aren't required to use `IHasColumn`, and neither required to use any of our other interfaces. But generally we expose such interfaces when they have a purpose, and I guess this gives it a purpose?

smoogipoo committed:
Merge branch 'master' into basic-compose-checks
